### DISCUSSION TITLE
Confused about distribution channels

### DISCUSSION THREAD

**Discussion by silouanwright** [semantic-release/semantic-release#2295] [Release workflow]
https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/release-workflow/distribution-channels.md

Right now, distribution channels are one of the more confusing features of semantic release. I get the whole "you merge it into a `beta` or `next`" and then you merge this into master, but really there's a lot of edge cases and considerations here right?

- **Experimental**: There's a lot of instances where I need to cut a release so that people can try it out in their projects and report back. It could be that I'm not necessarily done with that work. It could be that, based on the feedback they give me, I don't want to keep that work anymore. If we assume these are protected branches, what happens now? Do I have to revert these? Who's to say that it's too difficult with a big team that need to merge multiple "experimental" changes, that I'm going to be able to remember what needs to be reverted?
- Now you might say "Well you should accommodate this outside of semantic release" and I would definitely like to hear more about that. But my problem then, is that you're merging things into `next`, and then merging `master` into next. But there's no way to eject from that. If you realize after merging a couple of things into the next branch that you want to abandon some of that work... you can't, right?

So in essence, my 2 biggest questions around semantic release

1. Releasing experimental stuff
2. Walking back certain commits and changes, and how this looks in the workflow.

It would be really helpful to me if there could be more added to that workflow, because I don't think I'm totally getting it.

Last thing: I will say, before I configured semantic release, in my head the way I wanted this to work, is that there was some `beta` branch and any time master got updated, it would sync the `beta` branch up. That way I could push a bunch of work to check it out, and then when I push into master, the `beta` branch just resets itself, because it served it's purpose.

I don't know if this makes any sense but I'm happy to elaborate on it.

**Comment by travi** [comment]
the biggest thing to keep in mind that that semantic-release always operates from git history.

> the way I wanted this to work, is that there was some `beta` branch and any time master got updated, it would sync the `beta` branch up

in order for the history that is in the mainline branch to be in the history of the `beta` branch, and therefore influence how semantic-release operates on the `beta` branch, that mainline history must be merged into the `beta` branch. if you want this to be automated, managing the git operations is beyond the scope of semantic-release

> Walking back certain commits and changes

again, the answer here is to use git operations. you want to be careful to avoid modifying git history that already resulted in a release, so `git revert` is your friend. reverting previous changes that you want to abandon puts the history back in the state it was before the undesirable changes, which can then be released as a new version to follow the previous one.

> assume these are protected branches, what happens now?

pre-release branches are intended to be short-lived. they allow you to release changes on an alternative channel, like alpha or beta. these can be any non-default channel. the expectation is to branch from the latest mainline, release a series of versions while you harden features that are not yet ready for your stable channel, then merge the branch back to the mainline to promote these changes to the stable channel. after this promotion, the branch should be deleted and started new when there is a need for a new pre-release. if the experiment results in deciding that the changes should not be promoted, the branch can simply be abandoned and deleted

**Reply by silouanwright** [reply]
> if the experiment results in deciding that the changes should not be promoted, the branch can simply be abandoned and deleted

I understand what this means and the intention of it in theory, but I don't understand at all what this means in practice.

- Let's just assume, based on the default configuration, we're referring to the `beta` branch, which is a prerelease branch.
- You're recommending, when making changes to this branch (a remote branch), that when you want to discard it, you just simply delete the remote `beta` branch?
- What if there are 6 different developers, and they all want separate beta versions for the specific work they are completing?
- I feel like the best incarnation of this, would be to specify pre-release branches based on a regex pattern (branches starting with `beta-` or something, that way you're not working with some singular pre-release branch across developers that are merging random work in, and deleting the branch just so they can reset it.
- What happens if one dev merges work into meta, merges beta into master, forgets to delete beta, then more devs add work to master, then a dev merges something into beta, and merges that into master? Or better yet, a developer merges work into `beta`, decides they don't like the work, they forget about it, and another developer then pushes work into `beta`?

Again I don't if I'm missing something here but I am hilariously confused by how this is all supposed to work.

**Reply by travi** [reply]
> What if there are 6 different developers, and they all want separate beta versions for the specific work they are completing?

this point stands out to me. i think this might be pointing to part of the misunderstanding that could be resulting in the confusion you mention. can you help me understand the situation why you would have 6 different developers all working on unstable changes simultaneously? pre-releases are intended to be used for cases specific situations, like where you are trying to work through refining an api that you are unsure of the best shape that supports solving a need, or high risk situations like swapping bundling tools that are hard to ensure no breakages exist without testing in a consumer of the package. these situations should be rare, not a part of the team's common development workflow.

a fairly common misuse we see with pre-releases is teams attempting to emulate something like gitflow. semantic-release does not support that type of workflow. since you mention multiple simultaneous pre-releases, it sounds like you are likely trying to accomplish something other than gitflow. maybe you are wanting versions to be published so they can be QAed by testing WIP changes in a consumer of the package?

**Reply by silouanwright** [reply]
> maybe you are wanting versions to be published so they can be QAed by testing WIP changes in a consumer of the package?

100% correct. I'm working on a design system, and part of the way we validate changes, is to see them in action on our site. This would be somewhat easily done with `yarn link` but we want to push these changes out to an actual environment so that different non technical stakeholders can review them, and we can also QA them. We don't use gitflow workflows.

**Reply by travi** [reply]
Ok, that makes sense. However, the workflow you describe is not the intended use of pre-releases and is a workflow that we actively discourage. I recognize that linking is a common approach to accomplishing a similar goal, but that approach is also in the same category that we discourage. 

Instead, we strongly encourage enabling the ability to verify changes in isolation without requiring testing to happen in a consumer. For packages that include web ui aspects like you describe, I've had good luck with [storybook](https://storybook.js.org/) and [chromatic](https://www.chromatic.com/). With investment in the ability to verify your package independently, you can take pre-releases out of your normal workflow and consider pre-release for rare changes that are for more risky changes.

**Reply by silouanwright** [reply]
I understand the perspective there, and the reason you discourage that, but it is a common workflow. For our platform (and I think most platforms) you cannot fully anticipate all the ways your users will use your components. This is commonly why popular libraries will frequently break their consumers, because of some use case they did not anticipate. This is ok for an open source library: we just force users to file a ticket, and it gets put on the roadmap. In our use case, our library is for our platform, and breakage is not something we're willing to accommodate. So I do disagree, but I understand where you're coming from, and I think what you're suggesting is the ideal and primary use case. This also has to do with the way my company allocates QA resources as well currently. And we also do use Storybook as well.

In any case, whether it's the intended workflow or not, I will admit my goal is to publish betas, either on merge to beta, or on pull request to master. Is this something that you discourage, but can still be done within semantic release? Or is it just not possible? If not, are there any similar things out there that do this in an automated and not automated way, that you would suggest?

And again, apologies. I get what you mean, but this isn't something we can support at this time, for reasons I don't want to further make you suffer to read, and I'm now just trying to evaluate if the tool will assist me with this, or if I need to look elsewhere (and if so, if you have any suggestions).

**Reply by travi** [reply]
> This is ok for an open source library: we just force users to file a ticket, and it gets put on the roadmap. In our use case, our library is for our platform, and breakage is not something we're willing to accommodate.

do the consumers of your design system not use their normal QA process when pulling in an update? no matter what your QA process is for your design system, there is always a possibility that a bug will make its way through to your consumers. if found by your consumers, work needs to be done to fix that bug and a release must be made to make the fix available for your consumers. all i am suggesting is to verify your package enough in isolation to feel comfortable releasing "stable" releases rather than pre-releases. even doing pre-releases and extra manual verification likely does not prevent _all_ bugs from making it through, so your consumers still have to do due diligence before accepting an upgrade.

> I will admit my goal is to publish betas, either on merge to beta, or on pull request to master. Is this something that you discourage, but can still be done within semantic release? Or is it just not possible? If not, are there any similar things out there that do this in an automated and not automated way, that you would suggest?

using semantic-release in unsupported ways lands on your shoulders to determine. there is risk with such deviations that they may break in the future, even if you can get them working in the way you intend. you also will likely get no support from maintainers other than a conversation similar to this one, encouraging you to use it in supported ways.

**Reply by PawelWesolowski** [reply]
Branches `next` or `beta` are branches shared with others. Why shouldn't a feature get tested e2e (or with users) BEFORE merging it. Why should it be necessary to first merge and potentially destroy to check if sth got destroyed? 
Release cycle evolved in years from months to weeks, to days, to hours. E2E testing got also faster and delivery/deployments are almost instant. Building PR deployments should not be discouraged by this lib - it should be promoted.

**Comment by bombillazo** [comment]
Hello! I'm hijacking this convo, it has been very insightful!

Are distribution channels an external concept of semantic release? As in, the scope of a distribution channel is handled by an external process? Like, would NPM be considered a distribution channel?

How should semantic version work with these distribution channels per se?

**Reply by travi** [reply]
in the future, please open separate discussions for individual topics. 

to better understand the use of distribution channels within the npm ecosystem, please refer to [their documentation on the topic](https://docs.npmjs.com/cli/v8/commands/npm-dist-tag#purpose)
